Faster random.shuffle via static typing.

anntzer opened the pull request:
This patch modifies random.shuffle so that (when working on a ndarray) an array
of indices is shuffled and then elements are take()n from that array in that
order.  This allows the inner loop to be statically typed (it turns out this
is not so easy to write a generic shuffling code using Cython fused types) and
thus much faster (~6x for me), at the expense of a threefold increase in memory
use (I guess take() needs to create a copy, and an additional array of indices
is created.).

See #5514.

anntzer committed:
Faster random.shuffle via static typing.

This patch modifies random.shuffle so that (when working on a ndarray)
an array of indices is shuffled and then elements are take()n from that
array in that order.  This allows the inner loop to be statically typed
(it turns out this is not so easy to write a generic shuffling code
using Cython fused types) and thus much faster (~6x for me), at the
expense of a threefold increase in memory use (I guess take() needs to
create a copy, and an additional array of indices is created.).

See #5514.

njsmith commented:
To reduce the overhead in common cases, you could write a single loop for
64-bit integers, and then at the top level either call this directly (if
the input is int64), view-cast and then call this (if the input is e.g.
float64), or else allocate an index array and shuffle it.
On Dec 5, 2015 8:51 PM, "Antony Lee" <notifications@github.com> wrote:

> This patch modifies random.shuffle so that (when working on a ndarray) an
> array
> of indices is shuffled and then elements are take()n from that array in
> that
> order. This allows the inner loop to be statically typed (it turns out this
> is not so easy to write a generic shuffling code using Cython fused types)
> and
> thus much faster (~6x for me), at the expense of a threefold increase in
> memory
> use (I guess take() needs to create a copy, and an additional array of
> indices
> is created.).
>
> See #5514 <https://github.com/numpy/numpy/issues/5514>.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/numpy/numpy/pull/6776
> Commit Summary
>
>    - Faster random.shuffle via static typing.
>
> File Changes
>
>    - *M* numpy/random/mtrand/mtrand.pyx
>    <https://github.com/numpy/numpy/pull/6776/files#diff-0> (31)
>
> Patch Links:
>
>    - https://github.com/numpy/numpy/pull/6776.patch
>    - https://github.com/numpy/numpy/pull/6776.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6776>.
>

anntzer commented:
Indeed, this shaves off another 10% or so to the runtime (in the case where itemsize is the same as intp.itemsize).

Side question: currently, building numpy (via setup.py or runtests.py) fails (even after git clean -xf) with
```
Building, see build.log...
Running from numpy source directory.
Traceback (most recent call last):
  File "setup.py", line 263, in <module>
    setup_package()
  File "setup.py", line 247, in setup_package
    from numpy.distutils.core import setup
  File "/home/antony/src/numpy/numpy/distutils/__init__.py", line 21, in <module>
    from numpy.testing import Tester
  File "/home/antony/src/numpy/numpy/testing/__init__.py", line 12, in <module>
    from . import decorators as dec
  File "/home/antony/src/numpy/numpy/testing/decorators.py", line 21, in <module>
    from .utils import SkipTest
  File "/home/antony/src/numpy/numpy/testing/utils.py", line 18, in <module>
    from numpy.core import float32, empty, arange, array_repr, ndarray
  File "/home/antony/src/numpy/numpy/core/__init__.py", line 58, in <module>
    from numpy.testing import Tester
ImportError: cannot import name 'Tester'

Build failed!
```
which I can locally solve by first importing numpy.core in setup.py (to break the circular dependency), but is that a known issue?

anntzer committed:
Specialize shuffling for intp-sized 1d arrays.

This avoids the memory overhead of allocating an index array and
take()ing from the original array when possible.

seberg commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4652,33 +4652,33 @@ cdef class RandomState:
+                # Take from a shuffled range to benefit from static typing.
+                idxs = np.arange(n, dtype=np.intp)
+                self._shuffle_intpsized(idxs)
+                x.take(idxs, 0, out=x)
```
I doubt `np.take` does not work in-place or checks this. Which means the result can be corrupted by memory layout. I think you have to create a copy unfortunately.
Depends a bit on the shape, but unless your array is `(N, m)` with `m` being quite small, indexing is probably faster, but does not matter much.

jaimefrio replied:
If `mode` is the default `'raise'`, it always allocates an intermediate buffer, so that if an error in the indexing is detected, and an error has to be raised, the output array is left unchanged.  If you set it to `'clip'` or `'wrap'` then yes, you would start seeing all kind of funny things if you tried to do the operation in place.

seberg replied:
Ah ok, then take is nicer in that regard. Hopefully we have cipy on overlap at some point in indexing....


On Sun Dec 6 09:54:39 2015 GMT+0100, Jaime wrote:
> > -            with self.lock:
> > -                while i > 0:
> > -                    j = rk_interval(i, self.internal_state)
> > -                    buf[...] = x[j]
> > -                    x[j] = x[i]
> > -                    x[i] = buf
> > -                    i = i - 1
> > +        if isinstance(x, np.ndarray):
> > +            if x.ndim == 1 and x.dtype.itemsize == np.dtype(np.intp).itemsize:
> > +                # Directly shuffle the array if possible.
> > +                self._shuffle_intpsized(x.view(np.intp))
> > +            else:
> > +                # Take from a shuffled range to benefit from static typing.
> > +                idxs = np.arange(n, dtype=np.intp)
> > +                self._shuffle_intpsized(idxs)
> > +                x.take(idxs, 0, out=x)
> 
> If `mode` is the default `'raise'`, it always allocates an intermediate buffer, so that if an error in the indexing is detected, and an error has to be raised, the output array is left unchanged.  If you set it to `'clip'` or `'wrap'` then yes, you would start seeing all kind of funny things if you tried to do the operation in place.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/numpy/numpy/pull/6776/files#r4676812

anntzer commented:
There's a test failure on 2.6, due to MaskedArrays not supporting the buffer protocol there.  To be honest I don't really understand how the buffer protocol works on maskedarrays (how exactly is the mask information propagated?), moreover `buffer(MaskedArray(...))` works fine on 2.6.  I'll probably make the code conditional on the version of Python, given that 2.6 is expected to be dropped in a few versions.
Thoughts?

njsmith commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4652,33 +4652,33 @@ cdef class RandomState:
-            # independent of the array contents, so we can just swap directly.
             with self.lock:
-                while i > 0:
+                for i in reversed(range(1, n)):
```
Does cython know how to compile this down to a C for loop?

njsmith commented:
Actually I suspect that your implementation is broken for masked arrays in
all versions, because cython doesn't know about masks and will shuffle the
data array but not the masks. It would be good to add a test at a minimum.
On Dec 9, 2015 9:22 AM, "Antony Lee" <notifications@github.com> wrote:

> There's a test failure on 2.6, due to MaskedArrays not supporting the
> buffer protocol there. To be honest I don't really understand how the
> buffer protocol works on maskedarrays (how exactly is the mask information
> propagated?), moreover buffer(MaskedArray(...)) works fine on 2.6. I'll
> probably make the code conditional on the version of Python, given that 2.6
> is expected to be dropped in a few versions.
> Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6776#issuecomment-163330668>.
>

anntzer commented:
Indeed, the current test suite does not cover this case.  In fact the current implementation messes up the underlying array (i.e masking, shuffling and unmasking reveals that masked values have been overwritten by unmasked values), which should probably be considered a bug too.

Both issues are fixed by the new patch (which simply shuffle masked arrays by indexing with a shuffled index array).

anntzer committed:
Fix shuffling of masked arrays.

Also ensure that the masked data is preserved upon shuffling, which was
not the case before.

njsmith commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4652,33 +4652,35 @@ cdef class RandomState:
+        if isinstance(x, np.ndarray):
+            if (x.ndim == 1 and
+                x.dtype.itemsize == np.dtype(np.intp).itemsize and
+                not isinstance(x, np.ma.MaskedArray)):
```
I guess to be safer we might as well make it `type(x) is np.ndarray` to disallow all subclasses (since there's no guarantee that `MaskedArray` is the only weird subclass)

njsmith commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -4652,33 +4652,35 @@ cdef class RandomState:
                     x[i], x[j] = x[j], x[i]
-                    i = i - 1
+
+    cdef void _shuffle_intpsized(self, npy_intp[:] x):
```
Minor nit: should just be called `_shuffle_intps` I guess, since the itemsize check is happening elsewhere

njsmith commented:
I think that as written this code should be safe WRT not changing the random stream. But can you make sure that we have tests for that, that cover all the different branches, and add them if they're missing?

njsmith commented:
Aside from the minor nits above, the one thing I'm wary of is the memory overhead in the `take` path -- it's the sort of thing where we might well get users complaining that we totally broke their processing pipeline because they were shuffling some 10 gigabyte array and now... :-/.

One idea: would it be easy to use fused types to at least generate u8, u16, u32, u64 versions of `_shuffle_intps`, and use the view trick for other types that have exactly those sizes? then pretty much all the simple easy fast cases would be *really* fast, and we could stick to the old low-memory-overhead code for multidimensional and structured arrays.

anntzer commented:
This latest commit should handle all the issues you mentioned.  I added tests for various sizes of items.

anntzer committed:
shuffle: specialize std sizes; keep memory perf.

Do not rely on take(), which for non-standard sized arrays, thus
ensuing the previous memory performance at the expense of speed.  Also
get rid of the guarantee that shuffling masked arrays maintains the
masked values as well, which should probably be handled on np.ma's side
anyways.

ldoddema commented:
There should also be some special handling of object arrays, because these can't be viewed as int. You could branch on `dtype.hasobject`, but structured arrays with object members still pose a problem. These are bugged even in the current release:

```
np.random.shuffle(np.ones(3, 'i4, O'))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-38-67e839c634c1> in <module>()
----> 1 np.random.shuffle(np.ones(3, 'i4, O'))
mtrand.pyx in mtrand.RandomState.shuffle (numpy\random\mtrand\mtrand.c:30002)()
mtrand.pyx in mtrand.RandomState.shuffle (numpy\random\mtrand\mtrand.c:29939)()
ValueError: Setting void-array with object members using buffer.
```

anntzer commented:
FWIW, I realized that you can bypass the checks in `view` by using `np.frombuffer` instead.  A scary(ish) side note is that this opens a segfault possibility:
```
np.frombuffer(np.array(None), np.int64)[0] = 1
```
This handles the object array case but not the struct-with-object-fields case, which is arguably a bug in `ndarray.__setitem__` (and isn't a regression anyways):
```
In [23]: a = np.ones(3, 'i4, O')

In [24]: a[0] = a[1]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-5ef5a6077ce6> in <module>()
----> 1 a[0] = a[1]

ValueError: Setting void-array with object members using buffer.
```

homu commented:
:umbrella: The latest upstream changes (presumably #6910) made this pull request unmergeable. Please resolve the merge conflicts.

anntzer commented:
Superseded by #6933.
